domino/authentication.py

```diff
@@ -26,24 +26,29 @@ def __call__(self, r):
         return r
 
 
-def get_auth_by_type(api_key=None, auth_token=None, domino_token_file=None):
+def get_auth_by_type(api_key=None, auth_token=None, domino_token_file=None, api_key_from_env=None, domino_token_file_from_env=None):
     """
     Return appropriate authentication object for requests.
 
     If no authentication credential is provided, the call fails with an AssertError
 
     Precedence in the case of multiple credentials is:
-        1. auth token string
-        2. token file
-        3. API key
+        1. auth_token string
+        2. domino_token_file
+        3. api_key
+        4. domino_token_file_from_env
+        5. api_key_from_env
     """
 
-    assert any([api_key, auth_token, domino_token_file]), \
+    assert any([api_key, auth_token, domino_token_file, api_key_from_env, domino_token_file_from_env]), \
         "Unable to authenticate: no authentication method provided"
 
     if auth_token is not None:
         return BearerAuth(auth_token=auth_token)
     elif domino_token_file is not None:
         return BearerAuth(domino_token_file=domino_token_file)
-    else:
+    elif api_key is not None:
         return HTTPBasicAuth('', api_key)
+    else:
+        return self.get_auth_by_type(api_key=api_key_from_env, domino_token_file=domino_token_file_from_env)
```

Comment:
I feel like have a parameter called `something_from_env` is a code smell. Generally speaking, if something is in the environment, it doesn't have to be pass as a parameter. The fact that this method then somewhat unusually _called itself_ a second time under certain conditions made me think that this process should be simpler.

My suggestion is to simply encode the precedence all within this method. This seems like the most straightforward approach.
```suggestion
    if any([auth_token, domino_token_file, api_key]):
        if auth_token:
            return BearerAuth(auth_token=auth_token)
        elif domino_token_file:
            return BearerAuth(domino_token_file=domino_token_file)
        else:
            return HTTPBasicAuth('', api_key)
    else:
        # In the case that no authentication type was passed when this method
        # called, fall back to deriving the auth info from the environment.

        if os.getenv(DOMINO_TOKEN_FILE_KEY_NAME):
            return BearerAuth(domino_token_file=os.getenv(DOMINO_TOKEN_FILE_KEY_NAME))
        elif os.getenv(DOMINO_USER_API_KEY_KEY_NAME):
            return HTTPBasicAuth('', os.getenv(DOMINO_USER_API_KEY_KEY_NAME))
        else:
            # All attempts failed -- nothing to do but raise an error.
            raise RuntimeError("Unable to authenticate: no authentication provided")
```

Reply:
Note that you will need to change the first unit test as a result of this change.
```python
    with pytest.raises(AssertionError):
        Domino(host=dummy_host, project="anyuser/quick-start")
```
...will need to be
```python
    with pytest.raises(RuntimeError):
        Domino(host=dummy_host, project="anyuser/quick-start")
```